packages/neo-one-client-core/src/user/DapiUserAccountProvider.ts

```diff
@@ -0,0 +1,577 @@
+    return {
+      type: 'Boolean',
+      value,
+    };
+  }
+  if (value instanceof Buffer) {
+    return {
+      type: 'ByteArray',
+      value,
+    };
+  }
+  // tslint:disable-next-line strict-type-predicates
+  if (typeof value === 'object') {
+    return {
+      type: 'Array',
+      value: Object.entries(value).map(([k, v]) => [paramToArgDataType(k), paramToArgDataType(v)]),
+    };
+  }
+
+  /* istanbul ignore next */
+  throw new InvalidParamError(typeof value);
+};
+
+export class DapiUserAccountProvider<TProvider extends Provider> extends UserAccountProviderBase
+  implements UserAccountProvider {
+  public readonly currentUserAccount$: Observable<UserAccount | undefined>;
+  public readonly userAccounts$: Observable<ReadonlyArray<UserAccount>>;
+  public readonly networks$: Observable<ReadonlyArray<NetworkType>>;
+  private readonly dapi: Dapi;
+  private readonly initPromise: Promise<void>;
+  private mutableInitialized: boolean;
+  private readonly provider: TProvider;
+  private readonly currentAccountInternal$: BehaviorSubject<UserAccount | undefined>;
+  private readonly userAccountsInternal$: BehaviorSubject<ReadonlyArray<UserAccount>>;
+  private readonly networksInternal$: BehaviorSubject<ReadonlyArray<NetworkType>>;
+  private readonly defaultNetworkInternal$: BehaviorSubject<NetworkType | undefined>;
+
+  public constructor({ dapi, provider }: { readonly dapi: Dapi; readonly provider: TProvider }) {
+    super();
+    this.dapi = dapi;
+    this.provider = provider;
+    this.mutableInitialized = false;
+    this.initPromise = this.init();
+
+    this.userAccountsInternal$ = new BehaviorSubject<ReadonlyArray<UserAccount>>([]);
+    this.currentAccountInternal$ = new BehaviorSubject<UserAccount | undefined>(undefined);
+    this.networksInternal$ = new BehaviorSubject<ReadonlyArray<NetworkType>>([]);
+    this.defaultNetworkInternal$ = new BehaviorSubject<NetworkType | undefined>(undefined);
+
+    this.dapi.addEventListener('ACCOUNT_CHANGED', ({ address, label }) => {
+      this.dapi
+        .getPublicKey()
+        .then(({ publicKey }) => {
+          this.updateUserAccount({
+            network: this.defaultNetworkInternal$.getValue() as NetworkType,
+            publicKey,
+            address,
+            label,
+          });
+        })
+        .catch(() => {
+          // do nothing
+        });
+    });
+    this.dapi.addEventListener('NETWORK_CHANGED', (dapiNetworks) => {
+      this.updateNetworks(dapiNetworks);
+    });
+    this.dapi.addEventListener('DISCONNECTED', () => {
+      throw new Error(`The account connected to the dapp via the dapi provider has been disconnected (logged out).`);
+    });
+
+    this.userAccounts$ = this.userAccountsInternal$;
+    this.currentUserAccount$ = this.currentAccountInternal$.pipe(distinctUntilChanged());
+    this.networks$ = this.networksInternal$;
+  }
+
+  public getCurrentUserAccount(): UserAccount | undefined {
+    if (!this.mutableInitialized) {
```

Comment:
I don't think we need to check if it's initialized at all, seems like it's going to just throw an error that's going to be caught and ignore anyways.